WIP: qemu: NUMA Topology Reference Base

zvonkok opened the pull request:
For DGX or multi-numa systems with GPUs and NICs on both NUMA nodes we need to make sure we have the right NUMA topology in QEMU aka Kata.

koct9i committed:
qemu: enable NUMA support

Link qemu with libnuma and enable NUMA feature flag.

Signed-off-by: Konstantin Khlebnikov <koct9i@gmail.com>

koct9i committed:
kernel: enable x86-64 ACPI NUMA detection

CONFIG_NUMA is already on,
but without CONFIG_X86_64_ACPI_NUMA kernel cannot detect NUMA.

Signed-off-by: Konstantin Khlebnikov <koct9i@gmail.com>

koct9i committed:
runtime: add annotation default_maxmemory

It seems there is no annotation for disabling VM memory hotplug.
This should be useful at least until hotplug become fully NUMA-aware.

Signed-off-by: Konstantin Khlebnikov <koct9i@gmail.com>

koct9i committed:
runtime: add hypervisor options for NUMA topology

With enable_numa=true hypervisor will expose host NUMA topology as is:
map vm NUMA nodes to host 1:1 and bind vpus to relates CPUS.

Option "numa_mapping" allows to redefine NUMA nodes mapping:
- map each vm node to particular host node or several numa nodes
- emulate numa on host without numa (useful for tests)

Signed-off-by: Konstantin Khlebnikov <koct9i@gmail.com>

koct9i committed:
runtime: enforce NUMA topology by VCPU threads affinity

For optimal performance VCPU threads must utilize only NUMA local CPUs.

Signed-off-by: Konstantin Khlebnikov <koct9i@gmail.com>

koct9i committed:
govmm: setup qemu VM NUMA topology for initial CPUs and memory

If NUMA topology is enabled:
- groups CPUs into per NUMA node sockets
- split memory into per NUMA node modules
- report NUMA node for VCPU threads

Signed-off-by: Konstantin Khlebnikov <koct9i@gmail.com>

zvonkok commented:
Setting `enable_numa = true` in the config we get on a `0-1` NUMA system the following in the VM 
```
root@46e560ba43cc:/# cat /sys/devices/system/node/online
0-1
```

zvonkok commented:
Setting `default_vcpus = 16` we get in the VM 
```
root@ac7a3bf4e276:/# cat /sys/devices/system/node/online
0-1
root@ac7a3bf4e276:/# cat /sys/devices/system/node/node0/cpulist
0-15
root@ac7a3bf4e276:/#  cat /sys/devices/system/node/node1/cpulist

root@ac7a3bf4e276:/# 
```

And no, the host we have 
```
❯ ps -T -p 975075 -o pid,tid,comm
    PID     TID COMMAND
 975075  975075 qemu-system-x86
 975075  975078 qemu-system-x86
 975075  975080 IO mon_iothread
 975075  975081 CPU 0/KVM
 975075  975083 CPU 1/KVM
 975075  975084 CPU 2/KVM
 975075  975086 CPU 3/KVM
 975075  975087 CPU 4/KVM
 975075  975089 CPU 5/KVM
 975075  975090 CPU 6/KVM
 975075  975091 CPU 7/KVM
 975075  975092 CPU 8/KVM
 975075  975096 CPU 9/KVM
 975075  975097 CPU 10/KVM
 975075  975098 CPU 11/KVM
 975075  975099 CPU 12/KVM
 975075  975100 CPU 13/KVM
 975075  975101 CPU 14/KVM
 975075  975102 CPU 15/KVM
 975075  975104 vhost-975075
 975075  975635 vhost-975075
 975075  975637 vhost-975075
 975075  975638 vhost-975075
 975075  975639 vhost-975075
 975075  975640 vhost-975075
 975075  975641 vhost-975075
 975075  975642 vhost-975075
 975075  975643 vhost-975075
 975075  975644 vhost-975075
 975075  975645 vhost-975075
 975075  975646 vhost-975075
 975075  975647 vhost-975075
 975075  975648 vhost-975075
 975075  975649 vhost-975075
 975075  975650 vhost-975075
 975075  975651 vhost-975075
```
and 
```
❯ cat /sys/devices/system/node/node0/cpulist
0-55,112-167
```
Which means the first 56 CPUs are on node0 and correctly mapped in node0 in QEMU

zvonkok commented:
Now with 128 CPUs there is some mismatch in assigning the correct CPUs to the correct NUMA nodes.

zvonkok commented:
With the recent fixes (default_vcpus=16) we have now: 

```
1299991 1299999   1 CPU 0/KVM 0
1299991 1300005   2 CPU 1/KVM 0
1299991 1300006   7 CPU 2/KVM 0
1299991 1300007   4 CPU 3/KVM 0
1299991 1300008   5 CPU 4/KVM 0
1299991 1300009   0 CPU 5/KVM 0
1299991 1300010   3 CPU 6/KVM 0
1299991 1300011   6 CPU 7/KVM 0
1299991 1300012  61 CPU 8/KVM 1
1299991 1300013  56 CPU 9/KVM 1
1299991 1300014  59 CPU 10/KVM 1
1299991 1300015  62 CPU 11/KVM 1
1299991 1300016  57 CPU 12/KVM 1
1299991 1300017  60 CPU 13/KVM 1
1299991 1300018  63 CPU 14/KVM 1
1299991 1300019  58 CPU 15/KVM 1
```
8 CPUs pinned to node0 and 8 CPUs pinned to node1.

c3d commented on docs/how-to/how-to-set-sandbox-config-kata.md:
```diff
@@ -47,6 +47,7 @@ There are several kinds of Kata configurations and they are listed below.
 | `io.katacontainers.config.hypervisor.block_device_driver` | string | the driver to be used for block device, valid values are `virtio-blk`, `virtio-scsi`, `nvdimm`|
 | `io.katacontainers.config.hypervisor.cpu_features` | `string` | Comma-separated list of CPU features to pass to the CPU (QEMU) |
 | `io.katacontainers.config.hypervisor.default_max_vcpus` | uint32| the maximum number of vCPUs allocated for the VM by the hypervisor |
+| `io.katacontainers.config.hypervisor.default_maxmemory` | uint32| the maximum memory assigned for a VM by the hypervisor in `MiB` |
```
What is the use case? It's not really consistent with the commit message that talks about limiting hotplugging, not limiting the amount of memory the hypervisor takes.

I _think_ the idea is to ensure that you don't allocate beyond NUMA memory, but you should detail that a lot more in the commit message. What happens:
* If you add more memory than anticipated in the NUMA case?
* What does "hotplug become fully NUMA-aware" mean exactly? (What currently happens that is bad for Kata)

zvonkok replied:
Let me review the old commits. One issue with NUMA is that we cannot hot-plug memory.

fidencio replied:
My understanding (limited, very limited), is that memory and vcpu hotplug is quite complicated with NUMA, if do-able at all.

This is not really related to the content of this commit, but I wonder if we can add a logic to ensure that NUMA only works in case `static_sandbox_mgmt` is set to true, and then we avoid any kind of unexpected surprises for casual users.

Last but not least, we need to make sure that users will have a good message, info mesage, about the NUMA state and why it wasn't enabled on cases where they'd expect it to be.

zvonkok replied:
I think we should fail in this scenario and give a proper error message. I do not want to enable static resource mgmt implicitly.

fidencio replied:
I completely agree.

I also think a comment should be added in the config file saying something towards the lines of "if enabling numa support static_sandbox_resource_mgmt must also be enabled, as there's not hotplug support for NUMA"

c3d commented on src/runtime/pkg/oci/utils.go:
```diff
@@ -662,6 +662,16 @@ func addHypervisorMemoryOverrides(ocispec specs.Spec, sbConfig *vc.SandboxConfig
 	}
 
+	if err := newAnnotationConfiguration(ocispec, vcAnnotations.DefaultMaxMemory).setUintWithCheck(func(memorySz uint64) error {
+		if memorySz < vc.MinHypervisorMemory && sbConfig.HypervisorType != vc.RemoteHypervisor {
```
Why do you exclude the remote hypervisor case here?

zvonkok replied:
CSP VMs have the PCIe topology configured, but I guess for libvirt we could use it @stevenhorsman I do not have a lot of insights into the libvirt provider, would need to talk to Steve first if that makes sens.

c3d commented on src/runtime/virtcontainers/hypervisor_config_linux.go:
```diff
@@ -44,6 +44,10 @@ func validateHypervisorConfig(conf *HypervisorConfig) error {
 	}
 
+	if uint64(conf.MemorySize) > conf.DefaultMaxMemorySize {
+		conf.MemorySize = uint32(conf.DefaultMaxMemorySize)
```
Shouldn't you issue at least a warning in that case? Why is that not seen as a configuration error? And if anything, shouldn't things happen the other way round, if the objective is to just limit hotplug? The hotplug code compares the current memory size with the host memory size (which is where we take the max memory size into account IIUC). So if the memory size is already larger, should't that be respected?

Again, a bit more documentation in the commit messages about how you want to use this and what the intended effect is would help a lot.

zvonkok replied:
Yeah agree, more logging and docs would be worth while let me extend and elaborate on each fo the commits.

c3d commented on tools/packaging/kernel/configs/fragments/x86_64/acpi.conf:
```diff
@@ -4,6 +4,8 @@ CONFIG_X86_INTEL_PSTATE=y
 # Firecracker needs this to support `vcpu_count`
 CONFIG_X86_MPPARSE=y
 
+CONFIG_X86_64_ACPI_NUMA=y
```
Should that be the default for Kata? What is the cost in terms of kernel size / complexity?

zvonkok replied:
That is a fair point, let's make this configurable like ENABLE_NUMA and then only build for targets that explicitly require it

c3d commented on tools/packaging/scripts/configure-hypervisor.sh:
```diff
@@ -445,6 +445,9 @@ generate_qemu_options() {
 	qemu_options+=(functionality:--enable-seccomp)
 
+	# Support NUMA topology
+	qemu_options+=(functionality:--enable-numa)
```
Should that be the default? What is the cost in terms of qemu size? We have put a lot of efforts into trimming down qemu and the kernel for kata, so maybe adding back NUMA should only be done for those who need it?

For k8s, maybe this would end up as a different runtime class, kata-numa?

zvonkok replied:
No let's make this configurable as well that is tight to the kernel config.

c3d reviewed (changes requested):
Thanks @koct9i and @zvonkok.

I think this is an interesting development, but I would at least like to see a bit more documentation (if only in the commit messages) about the problems you are facing and how the new options address them.

Also, can you add measurements of the size increases in qemu and kernel for NUMA support. For scalability, we put a lot of effort into minimizing them, so if the change brings a lot of baggage that most people don't use, this may justify a separate build (e.g. different runtime class when using k8s) so that users who don't need NUMA don't pay the price.

zvonkok commented:
@c3d Thanks for the comments, I will work on a not that invasive NUMA support.

jaredquincy commented:
Is there a new separate PR for that "not that invasive NUMA support" thread @zvonkok ? I'd love to follow that if so.

fidencio commented:
> I think this is an interesting development, but I would at least like to see a bit more documentation (if only in the commit messages) about the problems you are facing and how the new options address them.

That's fair.

> Also, can you add measurements of the size increases in qemu and kernel for NUMA support. For scalability, we put a lot of effort into minimizing them, so if the change brings a lot of baggage that most people don't use, this may justify a separate build (e.g. different runtime class when using k8s) so that users who don't need NUMA don't pay the price.

This is not, though.  I understand downstream trying to keep this as tailored as possible to their usecase, but upstream we want to be generalist, as much as possible.

Right now I cannot run Kata Containers on any medium to large instance on CSPs without being hit by issues related to the lack of NUMA awareness of the project.

When we get to someone trying to use Kata Containers, we don't want the person to have to know what NUMA is or is not, to have to test with runtimeclass A, B, or C.  We just want the users to run the project and get the project running well, otherwise they'll turn their back to us and move to something easier to use.

With that said, I still can see an argument for having an option to disable NUMA during the build, but I think by default NUMA support should be enabled, providing a better out-of-the-box experience.

fidencio commented on src/runtime/virtcontainers/qemu_amd64.go:
```diff
@@ -196,7 +197,9 @@ func (q *qemuAmd64) cpuModel() string {
-	return genericMemoryTopology(memoryMb, hostMemoryMb, slots, q.memoryOffset)
+	memory := genericMemoryTopology(memoryMb, hostMemoryMb, slots, q.memoryOffset)
+	memory.MemoryModules = genericNUMAMemoryModles(memoryMb, 4, q.numaNodes)
+	return memory
```
cc @kevinzs2048 @BbolroC @Amulyam24 to check what should be done (if something should be done) for aarch64 / s390x / ppc64le.

zvonkok replied:
Ok than lets first enable it for x86 and arm64 and follow up with ppc64 if needed!

zvonkok commented on tools/packaging/kernel/configs/fragments/x86_64/acpi.conf:
```diff
@@ -4,6 +4,8 @@ CONFIG_X86_INTEL_PSTATE=y
 # Firecracker needs this to support `vcpu_count`
 CONFIG_X86_MPPARSE=y
 
+CONFIG_X86_64_ACPI_NUMA=y
```
This definitely works on Arm64. I developed the other PRs on a Grace Hopper machine. No kernel changes needed AFAIR

zvonkok committed:
runtime: Check NUMA config

If NUMA is enabled we also need to set static
resource management since hot-plug of CPU and memory
doe snot work with NUMA.

Signed-off-by: Zvonko Kaiser <zkaiser@nvidia.com>

zvonkok committed:
arm64: Add Memory Topology

Add proper NUMA memory topolgy support to arm64 qemu.

Signed-off-by: Zvonko Kaiser <zkaiser@nvidia.com>

fidencio commented:
cool, I'll go ahead and do the tests Tomorrow.

fidencio committed:
fixup! runtime: add hypervisor options for NUMA topology

fidencio committed:
fixup! kernel: enable x86-64 ACPI NUMA detection

fidencio committed:
fixup! govmm: setup qemu VM NUMA topology for initial CPUs and memory

fidencio commented on src/runtime/virtcontainers/qemu.go:
```diff
@@ -2669,6 +2669,36 @@ func genericMemoryTopology(memoryMb, hostMemoryMb uint64, slots uint8, memoryOff
+	memoryPerNode -= memoryPerNode % memoryAlign
+
+	// First NUMA node gets more if memory is not divide evenly.
+	moduleSize := memoryMb - memoryPerNode*uint64(len(numaNodes)-1)
```
So, basically what we're doing here is ensuring that:
moduleSize is memoryPerNode + the difference of memoryMb and what's been spread on the other nodes, correct?

This is not so easy to visualise (or I'm too tired).

fidencio commented on src/runtime/virtcontainers/qemu.go:
```diff
@@ -2669,6 +2669,36 @@ func genericMemoryTopology(memoryMb, hostMemoryMb uint64, slots uint8, memoryOff
+			HostNodes:    numaNode.HostNodes,
+			MemoryPolicy: "interleave",
+		})
+		moduleSize = memoryPerNode
```
The reason you set it here is because the first one can be bigger, but all the others should respect the memoryPerNode, correct?

fidencio commented on src/runtime/virtcontainers/qemu.go:
```diff
@@ -2669,6 +2669,36 @@ func genericMemoryTopology(memoryMb, hostMemoryMb uint64, slots uint8, memoryOff
+			MemoryPolicy: "interleave",
+		})
+		moduleSize = memoryPerNode
+		if moduleSize == 0 {
```
Do this before attributing the value, and check directly for memoryPerNode == 0

fidencio reviewed (commented):
I've done a first pass, created some fixups (pushed to your branch) and made some comments.

Static checks are failing as memory size has changed for some cases where NUMA is not enabled (I'll also investigate this, unless you do that before), and this is not really desired.

fidencio committed:
fixup! govmm: setup qemu VM NUMA topology for initial CPUs and memory

fidencio committed:
fixup! runtime: add annotation default_maxmemory

fidencio committed:
fixup! runtime: add hypervisor options for NUMA topology

fidencio commented:
We can close this one as it's been superseded by the NUMA PR that got merged.
